lib/Transforms/InsertGpuAllocs.cpp

```diff
@@ -0,0 +1,363 @@
+      } else {
+        op->emitError("Uhhandled mem op in gpu region");
+        return llvm::None;
+      }
+    };
+
+    auto scfDialect = getContext().getOrLoadDialect<mlir::scf::SCFDialect>();
+
+    auto hasMemAccess = [](mlir::Operation *op) -> bool {
+      if (auto memInterface =
+              mlir::dyn_cast<mlir::MemoryEffectOpInterface>(op)) {
+        if (memInterface.hasEffect<mlir::MemoryEffects::Read>() ||
+            memInterface.hasEffect<mlir::MemoryEffects::Write>())
+          return true;
+      }
+      if (auto call = mlir::dyn_cast<mlir::func::CallOp>(op)) {
+        for (auto arg : call.operands()) {
+          if (arg.getType().isa<mlir::MemRefType>())
+            return true;
+        }
+      }
+      return false;
+    };
+
+    // Traverse through all the memory access ops under GPU launch Op
+    // and add device memory allocation appropriately.
+    if (func.walk([&](mlir::Operation *op) {
+              if (!op->getParentOfType<mlir::gpu::LaunchOp>())
+                return mlir::WalkResult::advance();
+
+              if (!hasMemAccess(op))
+                return mlir::WalkResult::advance();
+
+              auto memref = getMemref(op);
+              if (!memref)
+                return mlir::WalkResult::interrupt();
+
+              for (auto mem : *memref) {
+                while (auto parentView =
+                           mem.getDefiningOp<mlir::ViewLikeOpInterface>())
+                  mem = parentView.getViewSource();
+
+                for (auto alias : aliases.resolve(mem)) {
+                  auto op = alias.getDefiningOp();
+                  if (op) {
+                    if (mlir::isa<mlir::memref::GetGlobalOp>(op)) {
+                      gpuGetMemrefGlobalParams.insert({op, {}});
+                      continue;
+                    }
+                    if (op->getDialect() == scfDialect ||
+                        mlir::isa<mlir::ViewLikeOpInterface>(op))
+                      continue;
+                    if (mlir::isa<mlir::memref::AllocOp>(op)) {
+                      gpuBufferAllocs.insert({op, {}});
+                    } else if (mlir::isa<mlir::func::CallOp>(op)) {
+                      // Ignore
+                    } else {
+                      op->emitError("Unhandled memref producer");
+                      return mlir::WalkResult::interrupt();
+                    }
+
+                  } else {
+                    auto block = alias.getParentBlock();
+                    auto blockArgs = block->getArguments();
+                    auto it = llvm::find(blockArgs, alias);
+                    assert(it != blockArgs.end());
+                    auto index = static_cast<unsigned>(it - blockArgs.begin());
+                    gpuBufferParams.insert({index, {}});
+                  }
+                }
+              }
+
+              return mlir::WalkResult::advance();
+            })
+            .wasInterrupted()) {
+      signalPassFailure();
```

Comment:
If the pass fail, will the L0runner fail?  Does GPU/L0 always guarantee that memrefs used inside kernel are always accessible, no matter where the data is located?

Reply:
yes, currently the runner will fail if the pass fails

Reply:
Does this pass work with Vulkan runner?

Reply:
no, vulkan runnner has different set of passes that it runs and we haven't tried this pass with vulkan...